circus-train-core/src/main/java/com/hotels/bdp/circustrain/core/replica/hive/AlterTableService.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (C) 2016-2020 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.core.replica.hive;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hotels.bdp.circustrain.core.replica.Replica;
+import com.hotels.hcommon.hive.metastore.client.api.CloseableMetaStoreClient;
+
+public class AlterTableService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Replica.class);
+
+  private CopyPartitionsOperation copyPartitionsOperation;
+  private RenameTableOperation renameTableOperation;
+
+  public AlterTableService(
+      CopyPartitionsOperation copyPartitionsOperation,
+      RenameTableOperation renameTableOperation) {
+    this.copyPartitionsOperation = copyPartitionsOperation;
+    this.renameTableOperation = renameTableOperation;
+  }
+
+  public void alterTable(CloseableMetaStoreClient client, Table oldTable, Table newTable) throws TException {
+    List<FieldSchema> oldColumns = oldTable.getSd().getCols();
+    List<FieldSchema> newColumns = newTable.getSd().getCols();
+    if (hasAnyChangedColumns(oldColumns, newColumns)) {
+      LOG
+          .info("Found columns that have changed type, attempting to recreate target table with the new columns."
+              + "Old columns: {}, new columns: {}", oldColumns, newColumns);
+      Table tempTable = new Table(newTable);
```

Comment:
Does this copy all the table properties?

Reply:
It does indeed. We can either (1) scrub all the table params here, or (2) I've created a `DropTableService` which removes custom table params before calling the drop table command. If thats overly complicated I can remove the new service and just do (1) instead.

Reply:
OK so if we removed the params here then we don't have to worry about the potential of Beekeeper picking up the drop of the temp table etc.? I think I prefer that approach unless some of the params are useful to have on the temp table? I don't think any of the intermediate operations use them in any way so probably not? And to be clear, the original table params will be kept on the final end result table?

Reply:
Okay cool, I'm happy to do this, I'll remove the DropTableService, and just remove the params as part of this service.

The intermediate ops dont use the params, but i think beekeeper may pick up on the alter partition events (though will ignore them as the location doesnt change). And yes, the original params will be on the final table - i'll add some tests for this.

Reply:
Tried this out, and actually it doesn't seem possible. We can't do an alter table on the table as it still has the old schema, so it fails with the same error. So the DropTableService looks better as we do the alter/drop once everything has been updated.

Reply:
OK, let's go with that for now and keep an eye on it when this is in use.